kotlinx-coroutines-core/jvm/test/RateLimiterTest.kt

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2016-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines
+
+import org.junit.Test
+import kotlin.test.*
+import kotlinx.coroutines.*
+
+internal class RateLimiterTest {
+
+    private val testDelays = listOf(1, 10, 100, 1000, 1200, 1990, 100_000)
+
+    @Test
+    internal fun `delaySequence sum is 1000`() {
+        val testResults = mutableListOf<Pair<Int, Long>>()
+        testDelays.forEach { eventsPerSec ->
+            val rateLimiter = RateLimiterImpl(eventsPerSec)
+            val seq = rateLimiter.createDelaySequence()
+            val delays = seq.take(eventsPerSec)
+            val fullSecond = delays.sum()
+            testResults.add(eventsPerSec to fullSecond)
+        }
+        testResults.forEach { (eventsPerSecond, fullRotationSum) ->
+            assertEquals(fullRotationSum, 1000L, "$eventsPerSecond events per second should sum to 1000 ms delay, but was $fullRotationSum")
+        }
+    }
+
+    @Test
+    internal fun `some known delays`() {
+        val testData = listOf<Pair<Int,Long>>(1 to 1000L, 1000 to 1L, 2 to 500L, 3 to 334L, 4 to 250L, 10_000 to 1L)
+        testData.forEach { (eventsPerSecond: Int, delayTime: Long) ->
+            val actual = RateLimiterImpl(eventsPerSecond).createDelaySequence().first()
+            assertEquals(actual, delayTime, "Calculated delay time $actual didn't match expected $delayTime for $eventsPerSecond events per second")
+        }
+    }
+
+    @Test(timeout = 2_000L)
+    internal fun `run for one second`(): Unit = runBlocking {
+        testDelays.map { eventsPerSec ->
+            launch {
+                val rateLimiter = rateLimiter(eventsPerSec)
+                val start = System.currentTimeMillis()
+                (0..eventsPerSec).forEach {
+                    rateLimiter.acquire()
```

Comment:
I think that to test that the rate-limiter really works as intended, it would be good to have a small delay inside the loop. That way the test will verify, that the delay introduced by your code depends on the amount of time taken between calls.

I'd say that ideally there are 2 tests, that test both scenarios (intervals between invocations < the delay interval, and > than the delay interval). That will cover both branches in the code and you can write stricter assertions by separating out the cases in different tests.

Reply:
Seems resonable
I was contemplating launching a bunch of jobs that all grabbed for permits, might achieve the same goal?

Reply:
I think that would be a good test but I'm not sure it will achieve the same goal of checking that delay varies with amount of work done in between calls.